File: app/src/main/java/org/autojs/autojs/core/looper/TimerThread.java
package org.autojs.autojs.core.looper;

import static org.autojs.autojs.util.StringUtils.str;

import android.os.Handler;
import android.os.Looper;

import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;

import org.autojs.autojs.concurrent.VolatileBox;
import org.autojs.autojs.engine.RhinoJavaScriptEngine;
import org.autojs.autojs.lang.ThreadCompat;
import org.autojs.autojs.runtime.ScriptRuntime;
import org.autojs.autojs.runtime.exception.ScriptInterruptedException;
import org.autojs.autojs6.R;

import java.util.concurrent.ConcurrentHashMap;

/**
 * Created by Stardust on 2017/12/27.
 */
public class TimerThread extends ThreadCompat {

    private static final ConcurrentHashMap<Thread, Timer> sTimerMap = new ConcurrentHashMap<>();

    private Timer mTimer;
    private final VolatileBox<Long> mMaxCallbackUptimeMillisForAllThreads;
    private final ScriptRuntime mRuntime;
    private final Runnable mTarget;
    private boolean mRunning = false;
    private final Object mRunningLock = new Object();

    public TimerThread(ScriptRuntime runtime, VolatileBox<Long> maxCallbackUptimeMillisForAllThreads, Runnable target) {
        super(target);
        mRuntime = runtime;
        mTarget = target;
        mMaxCallbackUptimeMillisForAllThreads = maxCallbackUptimeMillisForAllThreads;
    }

    @Override
    public void run() {
        mRuntime.loopers.prepare();
        mTimer = new Timer(mRuntime, mMaxCallbackUptimeMillisForAllThreads);
        sTimerMap.put(Thread.currentThread(), mTimer);
        ((RhinoJavaScriptEngine) mRuntime.engines.myEngine()).enterContext();
        notifyRunning();
        new Handler().post(mTarget);
        try {
            Looper.loop();
        } catch (Throwable e) {
            if (!ScriptInterruptedException.causedByInterrupted(e)) {
                mRuntime.console.error(Thread.currentThread() + ": ", e);
            }
        } finally {
            onExit();
            mTimer = null;
            org.mozilla.javascript.Context.exit();
            sTimerMap.remove(Thread.currentThread(), mTimer);
        }
    }

    @Override
    public void interrupt() {
        LooperHelper.quitForThread(this);
        super.interrupt();
    }

    private void notifyRunning() {
        synchronized (mRunningLock) {
            mRunning = true;
            mRunningLock.notifyAll();
        }
    }

    @CallSuper
    protected void onExit() {
        mRuntime.loopers.notifyThreadExit(this);
    }

    public static Timer getTimerForThread(Thread thread) {
        return sTimerMap.get(thread);
    }

    public static Timer getTimerForCurrentThread() {
        return getTimerForThread(Thread.currentThread());
    }

    public int setTimeout(Object callback, long delay, Object... args) {
        return getTimer().setTimeout(callback, delay, args);
    }

    public Timer getTimer() {
        if (mTimer == null) {
            throw new IllegalStateException(str(R.string.error_thread_is_not_alive));
        }
        return mTimer;
    }

    public boolean clearTimeout(int id) {
        return getTimer().clearTimeout(id);
    }

    public int setInterval(Object listener, long interval, Object... args) {
        return getTimer().setInterval(listener, interval, args);
    }

    public boolean clearInterval(int id) {
        return getTimer().clearInterval(id);
    }

    public int setImmediate(Object listener, Object... args) {
        return getTimer().setImmediate(listener, args);
    }

    public boolean clearImmediate(int id) {
        return getTimer().clearImmediate(id);
    }

    public void waitFor() throws InterruptedException {
        synchronized (mRunningLock) {
            if (!mRunning) {
                mRunningLock.wait();
            }
        }
    }

    @NonNull
    @Override
    public String toString() {
        return "Thread[" + getName() + "," + getPriority() + "]";
    }

}

File: app/src/main/java/org/autojs/autojs/core/looper/TimerThread.kt
package org.autojs.autojs.core.looper

import android.os.Handler
import android.os.Looper
import androidx.annotation.CallSuper
import org.autojs.autojs.AutoJs
import org.autojs.autojs.engine.RhinoJavaScriptEngine
import org.autojs.autojs.lang.ThreadCompat
import org.autojs.autojs.runtime.ScriptRuntime
import org.autojs.autojs.runtime.api.Console
import org.autojs.autojs.runtime.exception.ScriptInterruptedException
import org.mozilla.javascript.BaseFunction
import org.mozilla.javascript.Context
import java.lang.ref.WeakReference
import java.util.WeakHashMap
import java.util.concurrent.CountDownLatch

/**
 * Created by Stardust on Dec 27, 2017.
 * Modified by SuperMonster003 as of Jan 31, 2026.
 */
open class TimerThread(
    scriptRuntime: ScriptRuntime,
    private val target: Runnable,
) : ThreadCompat(target), ILooperThread {

    private val mRunningLatch = CountDownLatch(1)
    private var mWeakRuntime = WeakReference(scriptRuntime)

    @Volatile
    private var mTimer: Timer? = null

    @Volatile
    private var mLooper: Looper? = null

    override val looper: Looper?
        get() = mLooper

    override fun run() {
        val scriptRuntime = mWeakRuntime.get() ?: return

        scriptRuntime.loopers.prepare()

        scriptRuntime.timers.newTimer(scriptRuntime).also {
            mTimer = it
            sTimerMap[currentThread()] = WeakReference<Timer>(it)
        }

        val engine = scriptRuntime.engines.myEngine() as? RhinoJavaScriptEngine
        engine?.enterContext()

        notifyRunning()

        val currentLooper = Looper.myLooper()
        setLooper(currentLooper)
        currentLooper?.let { Handler(it).post(target) }

        val exceptionHandler: (t: Throwable) -> Unit = { t ->
            runCatching {
                if (!ScriptInterruptedException.causedByInterrupt(t)) {
                    val console: Console = mWeakRuntime.get()?.console ?: AutoJs.instance.globalConsole
                    console.error("${currentThread()}: $t")
                }
            }
        }

        runCatching { Looper.loop() }.onFailure(exceptionHandler)
        runCatching { onExit() }.onFailure(exceptionHandler)
        mTimer = null
        runCatching { Context.exit() }.onFailure(exceptionHandler)
        sTimerMap.remove(currentThread())
    }

    override fun interrupt() {
        LooperHelper.quit(LooperThread.getLooperOrNull(this))
        super.interrupt()
    }

    override fun toString() = "Thread[$name,$priority]"

    @CallSuper
    protected open fun onExit() {
        val loopers = mWeakRuntime.get()?.loopers ?: return
        loopers.notifyThreadExit(this)
    }

    @Throws(InterruptedException::class)
    fun waitFor() {
        mRunningLatch.await()
    }

    fun setTimeout(callback: BaseFunction): Double =
        setTimeout(callback, 1)

    fun setTimeout(callback: BaseFunction, delay: Long, vararg args: Any): Double = withTimer {
        setTimeout(callback, delay, args.copyOf())
    }

    fun clearTimeout(id: Double): Boolean = withTimer {
        clearTimeout(id)
    }

    fun setInterval(callback: BaseFunction): Double =
        setInterval(callback, 1L)

    fun setInterval(callback: BaseFunction, interval: Long, vararg args: Any): Double = withTimer {
        setInterval(callback, interval, args.copyOf())
    }

    fun clearInterval(id: Double): Boolean = withTimer {
        clearInterval(id)
    }

    fun setImmediate(callback: BaseFunction, vararg args: Any): Double = withTimer {
        setImmediate(callback, args.copyOf())
    }

    fun clearImmediate(id: Double): Boolean = withTimer {
        clearImmediate(id)
    }

    private inline fun <R> withTimer(callback: Timer.() -> R): R {
        val timer = mTimer
        checkNotNull(timer) { "Thread is not alive" }
        return callback(timer)
    }

    private fun setLooper(looper: Looper?) {
        mLooper = looper
    }

    private fun notifyRunning() {
        mRunningLatch.countDown()
    }

    companion object {

        private val sTimerMap = WeakHashMap<Thread, WeakReference<Timer>>()

        @JvmStatic
        fun getTimerForThread(thread: Thread) = sTimerMap[thread]?.get()
    }
}